mlflow/gluon.py

```diff
@@ -0,0 +1,214 @@
+import os
+
+import gorilla
+import mxnet as mx
+from mxnet import gluon
+from mxnet import sym
+from mxnet.gluon.contrib.estimator import Estimator, EpochEnd, TrainBegin, TrainEnd
+from mxnet.gluon.nn import HybridSequential
+
+import mlflow
+from mlflow import pyfunc
+from mlflow.exceptions import MlflowException
+from mlflow.models import Model
+from mlflow.tracking.artifact_utils import _download_artifact_from_uri
+from mlflow.utils import experimental
+from mlflow.utils.autologging_utils import try_mlflow_log
+
+FLAVOR_NAME = "gluon"
+_MODEL_SAVE_PATH = "net"
+
+
+@experimental
+def load_model(model_uri, ctx):
+    """
+    Load a Gluon model from a local file or a run.
+
+    :param model_uri: The location, in URI format, of the MLflow model. For example:
+
+                      - ``/Users/me/path/to/local/model``
+                      - ``relative/path/to/local/model``
+                      - ``s3://my_bucket/path/to/model``
+                      - ``runs:/<mlflow_run_id>/run-relative/path/to/model``
+
+                      For more information about supported URI schemes, see
+                      `Referencing Artifacts <https://www.mlflow.org/docs/latest/tracking.html#
+                      artifact-locations>`_.
+    :param ctx: Either CPU or GPU.
+
+    :return: A Gluon model instance.
+
+    >>> # Load persisted model as a Gluon model, make inferences against an NDArray
+    >>> model = mlflow.gluon.load_model("runs:/" + gluon_random_data_run.info.run_id + "/model")
+    >>> model(nd.array(np.random.rand(1000, 1, 32)))
+    """
+    local_model_path = _download_artifact_from_uri(artifact_uri=model_uri)
+
+    model_arch_path = os.path.join(local_model_path, "data", _MODEL_SAVE_PATH) + "-symbol.json"
+    model_params_path = os.path.join(local_model_path, "data", _MODEL_SAVE_PATH) + "-0000.params"
+    symbol = sym.load(model_arch_path)
+    inputs = sym.var('data', dtype='float32')
+    net = gluon.SymbolBlock(symbol, inputs)
+    net.collect_params().load(model_params_path, ctx)
+    return net
+
+
+class _GluonModelWrapper:
+    def __init__(self, gluon_model):
+        self.gluon_model = gluon_model
+
+    def predict(self, ndarray):
+        return self.gluon_model(ndarray)
+
+
+def _load_pyfunc(path):
+    """
+    Load PyFunc implementation. Called by ``pyfunc.load_pyfunc``.
+
+    :param path: Local filesystem path to the MLflow Model with the ``gluon`` flavor.
+    """
+    m = load_model(path, mx.current_context())
+    return _GluonModelWrapper(m)
+
+
+@experimental
+def save_model(gluon_model, path, mlflow_model=Model()):
```

Comment:
@cosmincatalin Can you add a `conda_env` parameter to the `save_model` method as well as a default conda environment that contains all of the dependencies necessary for using the gluon flavor? For an example, see the `get_default_conda_env()` method in the `mlflow.sklearn` module: https://github.com/mlflow/mlflow/blob/master/mlflow/sklearn.py#L39-L55.

This will ensure that all relevant dependencies are installed when users to serve or deploy a gluon model (e.g., via the `mlflow models serve` CLI).

Reply:
Done